src/librbd/ImageState.cc

```diff
@@ -23,6 +24,155 @@ using util::create_async_context_callback;
 using util::create_context_callback;
 
 template <typename I>
+class ImageUpdateWatchers {
+public:
+
+  ImageUpdateWatchers(I *image_ctx) : m_image_ctx(image_ctx),
+    m_lock(util::unique_lock_name("librbd::ImageUpdateWatchers::m_lock", this)) {
+  }
+
+  ~ImageUpdateWatchers() {
+    assert(m_watchers.empty());
+    assert(m_in_flight.empty());
+    assert(m_pending_unregister.empty());
+  }
+
+  void shut_down(Context *on_finish) {
+    CephContext *cct = m_image_ctx->cct;
+    ldout(cct, 20) << __func__ << dendl;
+    {
+      Mutex::Locker locker(m_lock);
+      assert(m_on_shut_down_finish == nullptr);
+      m_watchers.clear();
+      if (!m_in_flight.empty()) {
+	m_on_shut_down_finish = on_finish;
+	return;
+      }
+    }
+    ldout(cct, 20) << __func__ << ": completing shut down" << dendl;
+    on_finish->complete(0);
+  }
+
+  void register_watcher(UpdateWatchCtx *watcher, uint64_t *handle) {
+    CephContext *cct = m_image_ctx->cct;
+    ldout(cct, 20) << __func__ << ": watcher=" << watcher << dendl;
+
+    Mutex::Locker locker(m_lock);
+    assert(m_on_shut_down_finish == nullptr);
+
+    *handle = next_handle++;
+    m_watchers.insert(std::make_pair(*handle, watcher));
+  }
+
+  void unregister_watcher(uint64_t handle, Context *on_finish) {
+    CephContext *cct = m_image_ctx->cct;
+    ldout(cct, 20) << __func__ << ": handle=" << handle << dendl;
+
+    int r = 0;
+    {
+      Mutex::Locker locker(m_lock);
+      auto it = m_watchers.find(handle);
+      if (it == m_watchers.end()) {
+	r = -ENOENT;
+      } else {
+	if (m_in_flight.find(handle) != m_in_flight.end()) {
+	  assert(m_pending_unregister.find(handle) == m_pending_unregister.end());
+	  m_pending_unregister[handle] = on_finish;
```

Comment:
Perhaps instead of tracking pending unregisters, just erase the handle from ```m_watchers``` and at ```handle_notify```, if the handle is no longer part of ```m_watchers```, you know you were unregistered?